### ISSUE TITLE
Troubles with system command "file" and some foreign message (On Debian Jessie)

### ISSUE THREAD

**Issue by stefaweb** [mailwatch/MailWatch#976]
Hello!

I've a problem with some messages coming from Israel.

One of the problem seems related to MailWatch, the second to MailScanner.

First, in Message Viewer, the message is not well decoded.

![sc1](https://user-images.githubusercontent.com/12907102/32941903-ab99321a-cb87-11e7-84c2-e387b3eadfae.png)

![sc2](https://user-images.githubusercontent.com/12907102/32941947-cdbc6e98-cb87-11e7-8c90-2e11e0e0e31b.png)

But the message is well decoded in list view.

![unknown](https://user-images.githubusercontent.com/12907102/32941997-f03e25b0-cb87-11e7-8142-16183b1f720f.png)

Second, the message is blocked in red each time.

![sc3](https://user-images.githubusercontent.com/12907102/32942060-209c6e1a-cb88-11e7-90d7-db4909be0bbf.png)

The message have these inclosed files.

![sc5](https://user-images.githubusercontent.com/12907102/32942163-7db4f842-cb88-11e7-8572-c0fc61b877d4.png)

I don't understand why this message is blocked.

With Mailwatch 1.2.6 and MailScanner 5.0.6-5. Debian Jessie.

**Comment by stefaweb** [comment]
If needed I have the complete message as stored in Quarantine directory.

**Comment by stefaweb** [comment]
Hello!

Does anyone have an idea about this problem?

Cheers,

**Comment by shawniverson** [comment]
Not sure, but I am working on mailscanner stuff today, so shoot this email over to shawniverson@summitgrid.org and shawniverson@efa-project.org and I'll take a look.

**Comment by asuweb** [comment]
Does mailscanner log anything of interest when it processes this email?  Can you send the original directly to me?

**Comment by stefaweb** [comment]
Doing it. Thanks.

**Comment by asuweb** [comment]
Not much help, but it goes through fine on mine.  Can't replicate the issue with mailscanner blocking files.  It goes through fine.

The encoding of the subject on `status.php` is fine.  The encoding of the headers in `detail.php` is also correct.

The only issue I have is the headers in `viewmail.php` which aren't encoded correctly. The body of the message is fine though.

**Comment by stefaweb** [comment]
How do you inject it in your system?

I'll like to try the source mail on another server.

**Comment by stefaweb** [comment]
Sorry, one more.

Found this in MailScanner.conf:
`Attachment Encoding Charset = ISO-8859-1`

Seems the default MS value.

Is-it ok?

**Comment by asuweb** [comment]
That's what mine is set to.

To inject the message I do:

`sendmail -i -f 'from@somewhere.com' recipient@somewhere.com < file`

**Comment by stefaweb** [comment]
No way for me.

I've sent the mail to my email from one of my smtp server. 

The mail output from the smtp server and goes to the antispam server and was blocked again. Still wrong font printed in viewmail.php.

`MailScanner: No programs allowed (msg-9027-3.txt)`

`Warning: This message has had one or more attachments removed Warning: (msg-9027-3.txt). Warning: Please read the "ACTION-WEB-Attachment-Warning.txt" attachment(s) for more information. This is a message from the MailScanner E-Mail Virus Protection Service ---------------------------------------------------------------------- The original e-mail attachment "msg-9027-3.txt" is on the list of unacceptable attachments for this site and has been replaced by this warning message. If you wish to receive a copy of the original attachment, please e-mail helpdesk and include the whole of this message in your request. Alternatively, you can call them, with the contents of this message to hand when you call. At Wed Nov 22 15:55:15 2017 the virus scanner said: MailScanner: No programs allowed (msg-9027-3.txt) Note to Help Desk: Look on The XXXX (antispam-dev) MailScanner in /var/spool/MailScanner/quarantine/20171122 (message 1eHWQl-0004Zi-JK). -- Postmaster`

**Comment by stefaweb** [comment]
The complete sequence in logs.

```
Nov 22 15:55:15 antispam-dev MailScanner[9027]: New Batch: Scanning 1 messages, 270528 bytes
Nov 22 15:55:15 antispam-dev MailScanner[9027]: Filetype Checks: No executables (1eHWQl-0004Zi-JK msg-9027-3.txt)
Nov 22 15:55:15 antispam-dev MailScanner[9027]: Other Checks: Found 1 problems
Nov 22 15:55:15 antispam-dev MailScanner[9027]: Virus and Content Scanning: Starting
Nov 22 15:55:16 antispam-dev MailScanner[9027]: Virus Scanning completed at 319893 bytes per second
Nov 22 15:55:16 antispam-dev MailScanner[9027]: Saved entire message to /var/spool/MailScanner/quarantine/20171122/1eHWQl-0004Zi-JK
Nov 22 15:55:16 antispam-dev MailScanner[9027]: Saved infected "msg-9027-3.txt" to /var/spool/MailScanner/quarantine/20171122/1eHWQl-0004Zi-JK
Nov 22 15:55:16 antispam-dev MailScanner[9027]: Spam Checks: Starting
Nov 22 15:55:16 antispam-dev MailScanner[9027]: Expired 2 records from the SpamAssassin cache
Nov 22 15:55:16 antispam-dev MailScanner[9027]: MailWatch: Blacklist refresh time reached
Nov 22 15:55:16 antispam-dev MailScanner[9027]: MailWatch: Starting up MailWatch SQL Blacklist
Nov 22 15:55:16 antispam-dev MailScanner[9027]: MailWatch: Read 2853 blacklist entries
Nov 22 15:55:16 antispam-dev MailScanner[9027]: MailWatch: SQLNoScan refresh time reached
Nov 22 15:55:16 antispam-dev MailScanner[9027]: MailWatch: SQLSpamSettings:: Read 0 No Spam Scan entries
Nov 22 15:55:16 antispam-dev MailScanner[9027]: MailWatch: SQLSpamScores refresh time reached
Nov 22 15:55:16 antispam-dev MailScanner[9027]: MailWatch: SQLSpamSettings:: Read 3 Spam entries
Nov 22 15:55:16 antispam-dev MailScanner[9027]: MailWatch: SQLHighSpamScores refresh time reached
Nov 22 15:55:16 antispam-dev MailScanner[9027]: MailWatch: SQLSpamSettings:: Read 3 high Spam entries
Nov 22 15:55:19 antispam-dev MailScanner[9027]: Spam Checks completed at 95174 bytes per second
Nov 22 15:55:19 antispam-dev MailScanner[9027]: Cleaned: Delivered 1 cleaned messages
Nov 22 15:55:19 antispam-dev MailScanner[9027]: Deleted 1 messages from processing-database
Nov 22 15:55:19 antispam-dev MailScanner[9027]: Batch completed at 68381 bytes per second (270528 / 3)
Nov 22 15:55:19 antispam-dev MailScanner[9027]: Batch (1 message) processed in 3.96 seconds
Nov 22 15:55:19 antispam-dev MailScanner[9027]: MailWatch: Logging message 1eHWQl-0004Zi-JK to SQL
Nov 22 15:55:19 antispam-dev MailScanner[9027]: "Always Looked Up Last" took 0.00 seconds
Nov 22 15:55:19 antispam-dev MailScanner[9027]: Config: calling custom end function MailWatchLogging
Nov 22 15:55:19 antispam-dev MailScanner[9027]: Config: calling custom end function SQLNoScan
Nov 22 15:55:19 antispam-dev MailScanner[9027]: MailWatch: SQLSpamSettings:: Closing down MailWatch SQL No Scan
Nov 22 15:55:19 antispam-dev MailScanner[9027]: Config: calling custom end function SQLSpamScores
Nov 22 15:55:19 antispam-dev MailScanner[9027]: MailWatch: SQLSpamSettings:: Closing down MailWatch SQL Spam Scores
Nov 22 15:55:19 antispam-dev MailScanner[9027]: Config: calling custom end function SQLHighSpamScores
Nov 22 15:55:19 antispam-dev MailScanner[9027]: MailWatch: SQLSpamSettings:: Closing down MailWatch SQL High Spam Scores
Nov 22 15:55:19 antispam-dev MailScanner[9027]: Config: calling custom end function SQLBlacklist
Nov 22 15:55:19 antispam-dev MailScanner[9027]: MailWatch: Closing down MailWatch SQL Blacklist
Nov 22 15:55:19 antispam-dev MailScanner[9027]: MailScanner child dying of old age
Nov 22 15:55:19 antispam-dev MailScanner[15147]: MailWatch: 1eHWQl-0004Zi-JK: Logged to MailWatch SQL
```

**Comment by shawniverson** [comment]
I am concerned that this file does indeed look like an executable to the file command.  Will run some tests shortly...

**Comment by stefaweb** [comment]
The original message in Quarantine.

```
-rw-r--r-- 1 Debian-exim mtagroup 267215 nov.  17 10:44 /var/spool/MailScanner/quarantine/20171117/1eFdBq-0000b8-Ij/message`
-rw-r--r-- 1 Debian-exim mtagroup 7883 nov.  17 10:44 /var/spool/MailScanner/quarantine/20171117/1eFdBq-0000b8-Ij/msg-31454-44.txt
```

```
file -i /var/spool/MailScanner/quarantine/20171117/1eFdBq-0000b8-Ij/message
/var/spool/MailScanner/quarantine/20171117/1eFdBq-0000b8-Ij/message: message/rfc822; charset=us-ascii
file -i /var/spool/MailScanner/quarantine/20171117/1eFdBq-0000b8-Ij/msg-31454-44.txt
/var/spool/MailScanner/quarantine/20171117/1eFdBq-0000b8-Ij/msg-31454-44.txt: text/plain; charset=unknown-8bit
```

**Comment by shawniverson** [comment]
Yeah, I'm not running into the mailscanner issue either, although the viewmail.php is not rendering hebrew properly.

**Comment by shawniverson** [comment]
Ok, update on this issue, I can get the headers to render properly, but not the body.  I am going to prepare a PR for that.  It appears the quoted printable encoding of the email sample isn't valid...but I'm not sure...

**Comment by shawniverson** [comment]
Ok, another update,  the quoted printable is fine, but it is in ISO-8859-8.  I'm not sure how to tell viewmail.php to detect that and render that properly.  I can do  iconv('ISO-8859-8', 'UTF-8', $structure->body) .  But obviously that is not a fix.

**Comment by shawniverson** [comment]
Ok, got it!  Preparing another PR.

**Comment by stefaweb** [comment]
Hi @shawniverson!

I tested the two PRs. Works fine.

Concerning the problem with Mailscanner. Do you have some suggestion to look in my servers to try to identify the problem?

**Comment by stefaweb** [comment]
Sorry, still a problem in viewpart.php

![sc2](https://user-images.githubusercontent.com/12907102/33169249-389a7406-d044-11e7-9860-97ace0bbdbb5.jpg)

**Comment by shawniverson** [comment]
@stefaweb That part is actually rendered by viewmail.php.  Can you double check the application of PR #984 ?

**Comment by shawniverson** [comment]
@stefaweb Found a tiny error....rebased and pushed up to PR #984

**Comment by shawniverson** [comment]
@stefaweb concerning the no executables, I am baffled, because I get the exact same filetype output from the file command that you do, and we are both on the same version of mailscanner.  Can you verify that the file command is in fact the method used in MailScanner.conf?  It probably is, but I am pondering this....

**Comment by stefaweb** [comment]
I've this:

`File Command = /usr/bin/file`

Other thing to check?

**Comment by shawniverson** [comment]
@stefaweb You're not using postfix, right?

**Comment by stefaweb** [comment]
Yes.

To be sure,  I stopped mailscanner and sent the test message again. It arrived in exim spool directory.

```
ll /var/spool/exim4/input/1eHqkL-0008Bn-1q-*
-rw-r----- 1 Debian-exim Debian-exim 263417 nov.  23 13:36 /var/spool/exim4/input/1eHqkL-0008Bn-1q-D
-rw-r----- 1 Debian-exim Debian-exim   7111 nov.  23 13:36 /var/spool/exim4/input/1eHqkL-0008Bn-1q-H
```

**Comment by shawniverson** [comment]
@stefaweb I have a theory.  I don't think it is the actual message itself that is getting flagged.  The acompanying txt file that exim is generating for this particular email, however, appears to be getting flagged.  I'm not as familiar with exim.  Do you know what this file is for?

**Comment by stefaweb** [comment]
The text file seems generated by mailscanner. When the message arrive in exim spool, it contain only the header and the data (see below).

**Comment by shawniverson** [comment]
So mailscanner is writing that file.  Ok.  I have an idea.....can you grab a copy of the email from the exim spool before it enters mailscanner and email me it?  I want to send it through mine in a as pristine condition as possible.

**Comment by stefaweb** [comment]
ok done.

**Comment by stefaweb** [comment]
I've also found this.

When the received the message, a alter message from Mailscanner is enclosed to the message.

When read in Outlook, LF or CR are missing.

![sc3](https://user-images.githubusercontent.com/12907102/33174243-82589614-d057-11e7-83f3-340aeccdff1b.png)

But if I open the "source of mail" in Outlook all seems fine.

![sc4](https://user-images.githubusercontent.com/12907102/33174247-85d2ae42-d057-11e7-9482-23671f6ba47a.png)

**Comment by asuweb** [comment]
@stefaweb @shawniverson - Funnily enough, if you run the file `msg-31454-44.txt` through virustotal, it detects it as an DOS EXE.

(https://www.virustotal.com/#/file/74a12ba13e82497f5d045d82c9d88c1753f0f4aed513958533f4396f445eb9f8/details)

I think the issue is the way exim / mailscanner is generating that file.  On analysing the file with a text editor, it contains a plain text version of the original HTML one.

**Comment by shawniverson** [comment]
@asuweb My instance isn't creating that file.  So what part of mailscanner is generating it and why?

**Comment by stefaweb** [comment]
I have the art of falling on a funny problem. ;-)

**Comment by shawniverson** [comment]
@stefaweb based on the file you just sent me, I believe that the queue file format for exim appear different, which means a different part of mailscanner is probably parsing the queue file.  I am going to set up an exim instance and try to hack at this one.

**Comment by stefaweb** [comment]
@shawniverson and @asuweb: I've emailed to you my Mailscanner config.

**Comment by asuweb** [comment]
@shawniverson - neither is mine.

@stefaweb - yes, you win the prize for strange issues ;)

**Comment by stefaweb** [comment]
I changed in Mailscanner.conf:

Warning Is Attachment = yes (was no)

The alert message is now as attachment.

No change. Message still blocked.

**Comment by shawniverson** [comment]
Building a debian/exim/mailscanner :)

**Comment by shawniverson** [comment]
and testing the new 7zip support on debian too while I am at it

**Comment by stefaweb** [comment]
Configs. No change in /etc/MailScanner/MailScanner.conf.

/etc/MailScanner/conf.d/00MailWatch.conf
```
# MailScanner configuration file for Exim

Run As User = Debian-exim
Run As Group = Debian-exim

MTA = exim
Incoming Work User = Debian-exim
Incoming Work Group = mtagroup
Incoming Work Permissions = 0660

Detailed Spam Report = yes
Quarantine Whole Message = yes
Quarantine Whole Messages As Queue Files = no
Include Scores In SpamAssassin Report = yes
Quarantine User = Debian-exim
Quarantine Group = mtagroup
Quarantine Permissions = 0644

Always Looked Up Last = &MailWatchLogging
Is Definitely Not Spam = &SQLWhitelist
Is Definitely Spam = &SQLBlacklist

Incoming Queue Dir = /var/spool/exim4/input
Outgoing Queue Dir = /var/spool/exim4_outgoing/input

Sendmail = /usr/sbin/exim4
Sendmail2 = /usr/sbin/exim4 -DOUTGOING

SpamAssassin Local State Dir = /var/lib/spamassassin

%etc-dir% = /etc/MailScanner
%rules-dir% = /etc/MailScanner/rules
Filename Rules = %etc-dir%/filename.rules
Filetype Rules = %etc-dir%/filetype.rules
Dangerous Content Scanning = %rules-dir%/content.scanning.rules
Is Definitely Not Spam = %rules-dir%/spam.whitelist.rules

```

/etc/MailScanner/conf.d/server.conf

```
# MailScanner configuration file

%org-name% = xxx
%org-long-name% = 
%web-site% = xxx
Hostname = xxx
%report-dir% = /usr/share/MailScanner/reports/fr

Max Children = 10

Maximum Attachment Size = 30M
Deliver Unparsable TNEF = yes
Find UU-Encoded Files = yes

Virus Scanners = clamd
Virus Names Which Are Spam = Sane*UNOFFICIAL HTML/* *Phish* MBL*UNOFFICIAL *SecuriteInfo*UNOFFICIAL INetMsg.SpamDomain*UNOFFICIAL NPGX.DomainAddr*UNOFFICIAL NPGX.EmailAddr.*UNOFFICIAL winnow*UNOFFICIAL ScamNailer*UNOFFICIAL Bofhland*UNOFFICIAL Porcupine*UNOFFICIAL *Zip.Suspect*
Allow Password-Protected Archives = yes
Clamd Socket = /var/run/clamav/clamd.ctl
Clamd Lock File = /var/run/clamav/clamd.pid
Clamd Use Threads = no
Monitors for ClamAV Updates = /var/lib/clamav/*.cld /var/lib/clamav/*.cvd

Use SpamAssassin = &SQLNoScan
Required SpamAssassin Score = &SQLSpamScores
High SpamAssassin Score = &SQLHighSpamScores

Use Stricter Phishing Net = no
Quarantine Infections = yes
SpamScore Number Instead Of Stars = yes

Sign Clean Messages = no
Notify Senders = no
Notify Senders Of Blocked Filenames Or Filetypes = no
Notify Senders Of Other Blocked Content = no

Scanned Subject Text = {Scanné}
Virus Subject Text = {Virus ?}
Filename Subject Text = {Fichier ?}
Content Subject Text = {Contenu dangereux ?}
Size Subject Text = {Taille}
Disarmed Modify Subject = no
Disarmed Subject Text = {Script inactivé}
Phishing Subject Text = {Fraude ?}
Spam Subject Text = {Spam ?}
High Scoring Spam Subject Text = {Spam ?}
#Warning Is Attachment = no

Send Notices = no
Spam List = BARRACUDA SPAMHAUS SPAMCOP

Rebuild Bayes Every = 86400
Wait During Bayes Rebuild = yes

Spam Actions = store
High Scoring Spam Actions = delete
Bounce Spam As Attachment = yes
Max Spam Check Size = 900k

Log Speed = yes
Log Spam = yes

# To be loaded after %org-name%
Spam-Virus Header = X-%org-name%-MailScanner-SpamVirus-Report:
Mail Header = X-%org-name%-MailScanner:
Spam Header = X-%org-name%-MailScanner-SpamCheck:
Spam Score Header = X-%org-name%-MailScanner-SpamScore:
Information Header = X-%org-name%-MailScanner-Information:
Envelope From Header = X-%org-name%-MailScanner-From:
Envelope To Header = X-%org-name%-MailScanner-To:
ID Header = X-%org-name%-MailScanner-ID:
IP Protocol Version Header = # X-%org-name%-MailScanner-IP-Protocol:
Hostname = The %org-name% ($HOSTNAME) MailScanner
Attachment Warning Filename = %org-name%-Attachment-Warning.txt
Watermark Secret = %org-name%-Secret
Watermark Header = X-%org-name%-MailScanner-Watermark:
MCP Header = X-%org-name%-MailScanner-MCPCheck:

# End Of File
```

**Comment by shawniverson** [comment]
gotta put my exim hat on.....been a while....

**Comment by stefaweb** [comment]
Again one for Debian/Exim.

```
# To delete tmp files in SpamAssassin-Temp
@hourly /usr/bin/find /var/spool/MailScanner/incoming/SpamAssassin-Temp -name '.spamassassin*' -mmin +30 -print | xargs rm -rf

```

**Comment by shawniverson** [comment]
Ok, this stinks....I can't make it happen!  Playing with MailScanner configs now...

**Comment by shawniverson** [comment]
Not sure what I am missing....

Debian 9.2
Exim4 (in and out config based on https://www.mailscanner.info/exim/)
MailScanner 5.0.6-5
@stefaweb 's configuration

Ran both the message and the queue files through, nothing flagged as an exe

However, I do see some txt files generated temporarily in /var/spool/MailScanner/quarantine and are deleted upon completion of scanning, so I think they are work files, which I am guessing is normal.  Somehow I think one of the work files appears as an EXE type on @stefaweb 's setup, but not mine.

Need ideas...or notable differences (mine is not running in France.....I dunno.......)

**Comment by stefaweb** [comment]
No luck (for me) ☹️

I made some tests.

If I replace "File Command = /usr/bin/file -i" by "File Command = ", the message cross the gateway.

What is the risk to remove call to /usr/bin/file?

I've also looked at "No executables" and found theses occurrences.
(nothing in /usr/share/MailScanner/).

```
/etc/MailScanner/archives.filetype.rules.conf:deny	executable	No executables		No programs allowed
/etc/MailScanner/archives.filetype.rules.conf:deny	ELF		No executables		No programs allowed
/etc/MailScanner/filetype.rules.conf:deny	executable	No executables		No programs allowed
/etc/MailScanner/filetype.rules.conf:deny	ELF		No executables		No programs allowed

```

**Comment by shawniverson** [comment]
@stefaweb What version of this package do you have?

file    1:5.30-1+deb9u1

**Comment by stefaweb** [comment]
Hi @shawniverson!

file 1:5.22+15-2+deb8u3 from Jessie.

**Comment by stefaweb** [comment]
libmagic-mgc not installed.

**Comment by stefaweb** [comment]
@shawniverson 

What is result when we don't use "-i' with file on your server?

On mine.

```
file /var/spool/MailScanner/quarantine/20171123/1eHw2U-0005vq-N4/msg-22805-1.txt
/var/spool/MailScanner/quarantine/20171123/1eHw2U-0005vq-N4/msg-22805-1.txt: DOS executable (COM)

file -i /var/spool/MailScanner/quarantine/20171123/1eHw2U-0005vq-N4/msg-22805-1.txt
/var/spool/MailScanner/quarantine/20171123/1eHw2U-0005vq-N4/msg-22805-1.txt: text/plain; charset=unknown-8bit

file /var/spool/MailScanner/quarantine/20171123/1eHw2U-0005vq-N4/message
/var/spool/MailScanner/quarantine/20171123/1eHw2U-0005vq-N4/message: RFC 822 mail, ASCII text

file -i /var/spool/MailScanner/quarantine/20171123/1eHw2U-0005vq-N4/message
/var/spool/MailScanner/quarantine/20171123/1eHw2U-0005vq-N4/message: message/rfc822; charset=us-ascii
```

**Comment by stefaweb** [comment]
My be found something in MW code.

If we assume that the variable for "File Command" is $filecommand, I've found this line in SweepOther.pm:

`exec(map { m/(.*)/ } split(/ +/, $filecommand), @filelist);`

The "-i" is missing.

You have still in SweepOther.pm a second line with "-i" this time:

`exec(map { m/(.*)/ } split(/ +/, $filecommand), '-i', @filelist);`

**Comment by shawniverson** [comment]
Tada! 

Nov 24 06:09:51 efamail MailScanner[1295]: Filetype Checks: No executables (1eIBrh-0001rS-43 msg-1295-3.txt)
Nov 24 06:09:51 efamail MailScanner[1295]: Filetype Checks: No executables (1eIBrh-0001rR-3k msg-1295-1.txt)

Problem is specific to Debian Jessie and file 1:5.22+15-2+deb8u3.

The magic included in this package is not up to date, but there is no update for this package on Jessie...

**Comment by shawniverson** [comment]
Attempting to compile an up to date file/magic...https://github.com/file/file/

**Comment by shawniverson** [comment]
Compiled file from source did not resolve on Jessie....researching....

**Comment by shawniverson** [comment]
Fixed on Jessie, steps:

```
apt install build-essential libtool autoconf git
git clone  https://github.com/file/file/
cd file
autoreconf -f i
./configure
make
apt remove file libmagic1
make install
```
Edit /etc/MailScanner/MailScanner.conf:
```
File Command = /usr/local/bin/file
```

**Comment by stefaweb** [comment]
Great new @shawniverson!

Dependency problem.

```
# apt -d remove libmagic1
Lecture des listes de paquets... Fait
Construction de l'arbre des dépendances
Lecture des informations d'état... Fait
Certains paquets ne peuvent être installés. Ceci peut signifier
que vous avez demandé l'impossible, ou bien, si vous utilisez
la distribution unstable, que certains paquets n'ont pas encore
été créés ou ne sont pas sortis d'Incoming.
L'information suivante devrait vous aider à résoudre la situation :

Les paquets suivants contiennent des dépendances non satisfaites :
 libapache2-mod-php5 : Dépend: php5-cli mais ne sera pas installé
                       Dépend: libmagic1 mais ne sera pas installé
 php7.0-cli : Dépend: libmagic1 mais ne sera pas installé
E: Erreur, pkgProblem::Resolve a généré des ruptures, ce qui a pu être causé par les paquets devant être gardés en l'état.

# apt -d remove file
Lecture des listes de paquets... Fait
Construction de l'arbre des dépendances
Lecture des informations d'état... Fait
Les paquets suivants seront ENLEVÉS :
  debhelper equivs file gdebi-core libtool lintian
0 mis à jour, 0 nouvellement installés, 6 à enlever et 0 non mis à jour.
Après cette opération, 6 134 ko d'espace disque seront libérés.

```

**Comment by shawniverson** [comment]
Hmm, libapache2-mod-php5 depends on libmagic1.

```
dpkg -r --ignore-depends=libapache2-mod-php libmagic1
```
This should work around this and allow you to proceed...the magic installed by compiling file should allow libapache2-mod-php5 to continue to function.

**Comment by stefaweb** [comment]
Not good.

```
# dpkg -r --ignore-depends=libapache2-mod-php libmagic1
dpkg: un problème de dépendance empêche la suppression de libmagic1:amd64 :
 php5-cli dépend de libmagic1.
 file dépend de libmagic1 (= 1:5.22+15-2+deb8u3).
 python3-magic dépend de libmagic1 (>= 1:5.22+15-2+deb8u3).
 python3-magic dépend de libmagic1 (<< 1:5.22+15-2+deb8u3.1~).
 python3-magic dépend de libmagic1 (>= 1:5.22+15-2+deb8u3).
 python3-magic dépend de libmagic1 (<< 1:5.22+15-2+deb8u3.1~).
 libapache2-mod-php5 dépend de libmagic1 ; cependant :
  Le paquet libmagic1:amd64 doit être supprimé.

dpkg: erreur de traitement du paquet libmagic1:amd64 (--remove) :
 problème de dépendance - suppression ignorée
Des erreurs ont été rencontrées pendant l'exécution :
 libmagic1:amd64
```

**Comment by shawniverson** [comment]
Lol :)

```
dpkg -r --ignore-depends=libapache2-mod-php,php7.0-cli,file,python3-magic,php5-cli libmagic1
```

**Comment by stefaweb** [comment]
No end to this. Each time we add one, we got more.

# dpkg -r --ignore-depends=libapache2-mod-php5,file,python3-magic,php5-cli libmagic1

```
# apt -d remove file libmagic1
Lecture des listes de paquets... Fait
Construction de l'arbre des dépendances
Lecture des informations d'état... Fait
Le paquet « libmagic1 » n'est pas installé, et ne peut donc être supprimé
Vous pouvez lancer « apt-get -f install » pour corriger ces problèmes :
Les paquets suivants contiennent des dépendances non satisfaites :
 debhelper : Dépend: file (>= 3.23) mais ne sera pas installé
 gdebi-core : Dépend: file mais ne sera pas installé
 libapache2-mod-php5 : Dépend: libmagic1 mais ne sera pas installé
 libtool : Dépend: file mais ne sera pas installé
 lintian : Dépend: file mais ne sera pas installé
 php5-cli : Dépend: libmagic1 mais ne sera pas installé
 python3-magic : Dépend: libmagic1 (>= 1:5.22+15-2+deb8u3) mais ne sera pas installé
                 Dépend: libmagic1 (< 1:5.22+15-2+deb8u3.1~) mais ne sera pas installé
E: Dépendances non satisfaites. Essayez « apt-get -f install » sans paquet

```

**Comment by shawniverson** [comment]
Hahaha

I got another idea, let me see if I can package file and libmagic1 for you :)

**Comment by stefaweb** [comment]
I was looking for that. Build a backport is better.

**Comment by stefaweb** [comment]
How to revert back the `dpkg -r --ignore-depends`?

**Comment by stefaweb** [comment]
I'm sure that I'm not the only one to use Jessie.

**Comment by stefaweb** [comment]
Sorry. Solved with `apt-get -f install`.

**Comment by stefaweb** [comment]
Back on this.

Seems to work on my server if I just install the new "file" binary in /usr/local/bin and change` File Command = /usr/local/bin/file` (don't forget to restart mailscanner).

The message crossed the server without problem.

**Comment by shawniverson** [comment]
Debian Jessie Packages built for 5.32:

https://www.osstoolsofthetrade.org/file/file_5.32-1+deb8u3_amd64.deb
https://www.osstoolsofthetrade.org/file/file-dbg_5.32-1+deb8u3_amd64.deb
https://www.osstoolsofthetrade.org/file/libmagic1_5.32-1+deb8u3_amd64.deb
https://www.osstoolsofthetrade.org/file/libmagic-dev_5.32-1+deb8u3_amd64.deb
https://www.osstoolsofthetrade.org/file/python3-magic_5.32-1+deb8u3_all.deb
https://www.osstoolsofthetrade.org/file/python-magic_5.32-1+deb8u3_all.deb

**Comment by ghost** [comment]
Package build source, for the curious-minded.   Patches from 5.22 are resolved for 5.32.

https://osstoolsofthetrade.org/file/file_5.32.orig.tar.gz

**Comment by stefaweb** [comment]
Great @shawniverson!

**Comment by stefaweb** [comment]
No yet tried but at first seems ok from apt point of view.

```
e# apt-get -d upgrade
Lecture des listes de paquets... Fait
Construction de l'arbre des dépendances
Lecture des informations d'état... Fait
Calcul de la mise à jour... Fait
Les paquets suivants seront mis à jour :
  file libmagic1 python3-magic
3 mis à jour, 0 nouvellement installés, 0 à enlever et 0 non mis à jour.
Il est nécessaire de prendre 406 ko dans les archives.
Après cette opération, 2 152 ko d'espace disque supplémentaires seront utilisés.
Souhaitez-vous continuer ? [O/n]
ATTENTION : les paquets suivants n'ont pas été authentifiés.
  python3-magic file libmagic1
Faut-il installer ces paquets sans vérification ? [o/N] o
Réception de : 1 http://deb.xxxx.fr/debian/ jessie/main python3-magic all 1:5.32-1+deb8u3 [47,6 kB]
Réception de : 2 http://deb.xxxx.fr/debian/ jessie/main file amd64 1:5.32-1+deb8u3 [63,5 kB]
Réception de : 3 http://deb.xxxx.fr/debian/ jessie/main libmagic1 amd64 1:5.32-1+deb8u3 [295 kB]
406 ko réceptionnés en 0s (15,9 Mo/s)
Téléchargement achevé et dans le mode téléchargement uniquement
```

**Comment by stefaweb** [comment]
Works!

```
# file /var/spool/MailScanner/quarantine/20171123/1eHw2U-0005vq-N4/msg-22805-1.txt
/var/spool/MailScanner/quarantine/20171123/1eHw2U-0005vq-N4/msg-22805-1.txt: Non-ISO extended-ASCII text

```

And the message crosse the gateway.

**Comment by stefaweb** [comment]
Hello!

Seems the problem is still present with Debian Stretch and file 5.30. No backport available.

Cheers,

**Comment by stefaweb** [comment]
@shawniverson !

The problem is still present with Stretch. No backport exist.

Would it be possible to make debian packages for Stretch using the version 5.35 of binary file in sid or buster with the associated library as libmagic1 and python3-magic ?

**Comment by ghost** [comment]
@stefaweb Sorry for late reply, been sick with the flu.  Should be doable. I'll give it a shot.

**Comment by ghost** [comment]
@stefaweb 

https://osstoolsofthetrade.org/file/file-5.36-packages.tar.gz

**Comment by stefaweb** [comment]
Thanks @shawniverson.

I will try it and let you know.

**Comment by stefaweb** [comment]
@shawniverson

Problem with the package. The version I got is 5.30.

I installed the packages using my private deb repository.

`apt-get install python-magic python3-magic libmagic1-dbgsym file-dbgsym libmagic-mgc libmagic-dev file`

`ii  file                                  1:5.36-1+deb9u2                    amd64        Recognize the type of data in a file using "magic" numbers
`

```
# /usr/bin/file -v
file-5.30
```

**Comment by Skywalker-11** [comment]
Did you check that the repo is actually recognized as valid package source/is prioritized with eg. `apt policy file`

**Comment by stefaweb** [comment]
Hi @Skywalker-11!

I'm sure. Using dpkg directly done the same.

```
# dpkg -i file_5.36-1+deb9u2_amd64.deb
(Reading database ... 66193 files and directories currently installed.)
Preparing to unpack file_5.36-1+deb9u2_amd64.deb ...
Unpacking file (1:5.36-1+deb9u2) over (1:5.36-1+deb9u2) ...
Setting up file (1:5.36-1+deb9u2) ...
Processing triggers for man-db (2.7.6.1-2) ...
# /usr/bin/file -v
file-5.30
magic file from /usr/local/share/misc/magic
```

**Comment by ghost** [comment]
@stefaweb I must have packaged the wrong version of the binaries.  I will rebuild the packages.

**Comment by ghost** [comment]
@stefaweb back from vacation, working on repackaging

**Comment by ghost** [comment]
@stefaweb Repackaged, looks good now.

https://osstoolsofthetrade.org/file/file-5.36-packages.tar.gz

**Comment by stefaweb** [comment]
@shawniverson 

Thanks a lot. I'll try and let you know.

**Comment by stefaweb** [comment]
@shawniverson 

Seems we still have a trouble.

The regular file binary on Stretch:

```
root@local:/home/sleclerc# file -v
file-5.30
magic file from /etc/magic:/usr/share/misc/magic
```

Your packaged file version:

```
root@antispam:/home/sleclerc# file -v
file-5.36
magic file from /usr/local/share/misc/magic
```

`/etc/magic` is missing.

**Comment by ghost** [comment]
@stefaweb But does it work?

**Comment by stefaweb** [comment]
For that, I've to wait to received the right mail with the infamous .dat. ;)

**Comment by Skywalker-11** [comment]
> 
> `/etc/magic` is missing.

The default /etc/magic file only contains comments and is only used for local customizations (eg. custom file formats) which usually are not necessary.

**Comment by stefaweb** [comment]
Hello!

I received a mail with the .dat file. The new binary file seem to work fine now.

```
# file -i /var/spool/MailScanner/quarantine/20190415/28E51BC0053.AEEE6/740000.dat
/var/spool/MailScanner/quarantine/20190415/28E51BC0053.AEEE6/740000.dat: application/x-dosdriver; charset=binary
```

```
# file /var/spool/MailScanner/quarantine/20190415/28E51BC0053.AEEE6/740000.dat
/var/spool/MailScanner/quarantine/20190415/28E51BC0053.AEEE6/740000.dat: DOS executable (block device driver)
```

But, the .dat file are still blocked.

```
MailScanner: No programs allowed (740000.dat)
```

I've this in /etc/MailScanner/filename.rules.conf

`allow   \.dat$`

Starting to be crazy this story !

**Comment by stefaweb** [comment]
Hello!

I'm back on this. I made some test with a mail I know causing problem.

sendmail -i -f 'from@domain.tld' to@domain.tld < message_with_dat

In archives.filetype.rules.conf, I have:

```
allow   DOS executable (block device driver)    -                       -
allow   x-dosdriver     -                       -
deny    self-extract    No self-extracting archives     No self-extracting archives allowed
deny    executable      No executables          No programs allowed
```

in archives.filename.rules.conf, I have:

`allow   \.dat$                  -       -`

With this, the message is blocked by MailScanner instead.

```
Apr 19 10:02:57 antispam MailScanner[20325]: Saved infected "390000.dat" to /var/spool/MailScanner/quarantine/20190419/D25AFBC069D.AC2A1
Apr 19 10:02:57 antispam MailScanner[20325]: Saved infected "740000.dat" to /var/spool/MailScanner/quarantine/20190419/D25AFBC069D.AC2A1
Apr 19 10:02:57 antispam MailScanner[20325]: Saved infected "130000.dat" to /var/spool/MailScanner/quarantine/20190419/D25AFBC069D.AC2A1
```

```
Apr 19 10:02:40 antispam MailScanner[20325]: Filetype Checks: No executables (D25AFBC069D.AC2A1 0000.dat)
Apr 19 10:02:40 antispam MailScanner[20325]: Filetype Checks: No executables (D25AFBC069D.AC2A1 0000.dat)
Apr 19 10:02:40 antispam MailScanner[20325]: Filetype Checks: No executables (D25AFBC069D.AC2A1 0000.dat)
Apr 19 10:02:57 antispam MailScanner[20325]: Saved entire message to /var/spool/MailScanner/quarantine/20190419/D25AFBC069D.AC2A1
Apr 19 10:02:57 antispam MailScanner[20325]: Saved infected "3AE_2949.docx" to /var/spool/MailScanner/quarantine/20190419/D25AFBC069D.AC2A1
Apr 19 10:02:57 antispam MailScanner[20325]: Saved infected "390000.dat" to /var/spool/MailScanner/quarantine/20190419/D25AFBC069D.AC2A1
Apr 19 10:02:57 antispam MailScanner[20325]: Saved infected "740000.dat" to /var/spool/MailScanner/quarantine/20190419/D25AFBC069D.AC2A1
Apr 19 10:02:57 antispam MailScanner[20325]: Saved infected "130000.dat" to /var/spool/MailScanner/quarantine/20190419/D25AFBC069D.AC2A1
Apr 19 10:02:57 antispam MailScanner[20325]: Saved infected "13RC_2949_ANNEXESFINANCIERES.xlsx" to /var/spool/MailScanner/quarantine/20190419/D25AFBC069D.AC2A1
Apr 19 10:02:57 antispam MailScanner[20325]: Saved infected "4AE_2949_A2_ANNEXESFINANCIERES.xlsx" to /var/spool/MailScanner/quarantine/20190419/D25AFBC069D.AC2A1
Apr 19 10:02:57 antispam MailScanner[20325]: Message D25AFBC069D.AC2A1 from xx.xx.xx.xx (to@domain.tld) to domain.tld is too big for spam checks (11227855 > 1024000 bytes)
Apr 19 10:03:00 antispam MailScanner[20325]: Requeue: D25AFBC069D.AC2A1 to 861D8BC0B36
Apr 19 10:03:00 antispam MailScanner[20325]: MailWatch: Logging message D25AFBC069D.AC2A1 to SQL
```

One of the .dat file:

```
# file /var/spool/MailScanner/quarantine/20190419/D25AFBC069D.AC2A1/740000.dat
/var/spool/MailScanner/quarantine/20190419/D25AFBC069D.AC2A1/740000.dat: DOS executable (block device driver)
```

```
# file -i /var/spool/MailScanner/quarantine/20190419/D25AFBC069D.AC2A1/740000.dat
/var/spool/MailScanner/quarantine/20190419/D25AFBC069D.AC2A1/740000.dat: application/x-dosdriver; charset=binary
```

If I comment this line:

`deny    executable      No executables          No programs allowed
`

The message cross MailScanner.

If someone have an idea.

**Comment by ghost** [comment]
@stefaweb If the latest version of file magic still flags this file as an executable, we have a problem that I cannot fix downstream.  You may need to engage the maintainers of the file command.  If this dat file has the exact same preamble as an executable, it may be impossible to distinguish from a real executable.  A hex editor would be needed to confirm that.  If that is the case, your only recourse will be to allow executables or create a ruleset for certain domains to allow them.

**Comment by stefaweb** [comment]
Hi @shawniverson!

Back from short vacations.

The .dat file seems "empty" with only 00000. The all start with FFFF FFFF.

`# xxd /var/spool/MailScanner/quarantine/20190419/D25AFBC069D.AC2A1/130000.dat
`
```
00000000: ffff ffff 0000 0000 0000 0000 0000 0000  ................
00000010: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000020: 0000 0000 0000 0000 0000 0000 0000 0000  ................
```

```
file -i /var/spool/MailScanner/quarantine/20190419/D25AFBC069D.AC2A1/130000.dat
/var/spool/MailScanner/quarantine/20190419/D25AFBC069D.AC2A1/130000.dat: application/x-dosdriver; charset=binary
```

`-rw-rw-r-- 1 postfix mtagroup 2797 Apr 19 10:02 /var/spool/MailScanner/quarantine/20190419/D25AFBC069D.AC2A1/740000.dat
`

I have a look on different .dat files matched as executable. All are like my sample.

**Comment by Skywalker-11** [comment]
This seems to be correct as it is the magic number of "DOS system driver" per [https://www.garykessler.net/library/file_sigs.html](https://www.garykessler.net/library/file_sigs.html) and [https://filesignatures.net/index.php?search=ffffffff&mode=SIG](https://filesignatures.net/index.php?search=ffffffff&mode=SIG)
